src/compiler/checker.ts

```diff
@@ -10506,6 +10887,31 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 return localName;
             }
         }
+
+        function isUnfolding(context: NodeBuilderContext) {
+            return context.unfoldDepth !== -1;
+        }
+
+        function isHashPrivate(s: Symbol): boolean {
+            // `valueDeclaration` could be undefined if inherited from
+            // a union/intersection base type, but inherited properties
+            // don't matter here.
+            return !!s.valueDeclaration && isNamedDeclaration(s.valueDeclaration) && isPrivateIdentifier(s.valueDeclaration.name);
+        }
+
+        function getHashPrivateName(s: Symbol): PrivateIdentifier | undefined {
+            if (s.valueDeclaration && isNamedDeclaration(s.valueDeclaration) && isPrivateIdentifier(s.valueDeclaration.name)) {
+                return factory.cloneNode(s.valueDeclaration.name);
+            }
+            return undefined;
+        }
+    }
+
+    /** Returns true if a type is declared in a lib file. */
+    // Don't unfold types like `Array` or `Promise`, instead treating them as transparent.

```

Comment:
nit: 
```suggestion
    // Don't unfold types like `Array` or `Promise`, instead treating them as opaque.

```
If they were transparent, we'd be looking through them, I think